stremio-core-protobuf/src/bridge/loadable.rs

```diff
@@ -1,33 +1,45 @@
-use stremio_core::deep_links::MetaItemDeepLinks;
-use stremio_core::models::common::{Loadable, ResourceError};
-use stremio_core::models::ctx::{Ctx, CtxError};
-use stremio_core::models::link::LinkError;
-use stremio_core::models::streaming_server::PlaybackDevice;
-use stremio_core::runtime::EnvError;
-use stremio_core::types::addon::{Descriptor, DescriptorPreview, ResourcePath, ResourceRequest};
-use stremio_core::types::api::{
-    GetModalResponse, GetNotificationResponse, LinkAuthKey, LinkCodeResponse,
-};
-use stremio_core::types::library::LibraryItem;
-use stremio_core::types::resource::{MetaItem, MetaItemPreview, Stream, Subtitles};
-use stremio_core::types::streaming_server::{Settings, Statistics};
-use stremio_watched_bitfield::WatchedBitField;
+use itertools::Itertools;
 use url::Url;
 
-use crate::bridge::ToProtobuf;
-use crate::protobuf::stremio::core::models;
-use crate::protobuf::stremio::core::models::{LoadedModal, LoadedNotification, PlaybackDevices};
+use stremio_core::{
+    deep_links::MetaItemDeepLinks,
+    models::{
+        common::{Loadable, ResourceError},
+        ctx::{Ctx, CtxError},
+        link::LinkError,
+        streaming_server::PlaybackDevice,
+    },
+    runtime::EnvError,
+    types::{
+        addon::{Descriptor, DescriptorPreview, ResourcePath, ResourceRequest},
+        api::{GetModalResponse, GetNotificationResponse, LinkAuthKey, LinkCodeResponse},
+        library::LibraryItem,
+        resource::{MetaItem, MetaItemPreview, Stream, Subtitles},
+        streaming_server::{Settings, Statistics},
+        watched_bitfield::WatchedBitField,
+    },
+};
+
+use crate::{
+    bridge::ToProtobuf,
+    protobuf::stremio::core::models::{self, LoadedModal, LoadedNotification, PlaybackDevices},
+};
 
 impl ToProtobuf<models::loadable_page::Content, (&Ctx, &ResourceRequest)>
     for Loadable<Vec<MetaItemPreview>, ResourceError>
 {
-    fn to_protobuf(
+    fn to_protobuf<E: stremio_core::runtime::Env + 'static>(
         &self,
         (ctx, request): &(&Ctx, &ResourceRequest),
     ) -> models::loadable_page::Content {
         match &self {
             Loadable::Ready(ready) => models::loadable_page::Content::Ready(models::Page {
-                meta_items: ready.to_protobuf(&(*ctx, *request)),
+                meta_items: ready
+                    .iter()
+                    .unique_by(|meta_item| &meta_item.id)
```

Comment:
Deduping here is not a great idea IMO, as it can mess up addon pagination and when the UI might not know that it needs to load the next page (since items were removed indexes of those items change). So IO it should be done on the UI side where the UI dedups, but the model holder holds unmodified list.

Reply:
Sadly this is already the case for many parts of the stremio-core-web. We do a lot of the UI processing there.

I was working on unifying the logic but it will need a bit more work due to differences in structures used in the UIs (core-web vs others).
I'd rather keep it all in the same place (the stremio-core-*) than splitting it into different parts of the UIs for the different platforms.

Reply:
Hmm, but I don't think we should add buggy and erroneous logic, just because it's done in web? That's not a good reason :sweat_smile:

Reply:
This is not a good example of the problem we are fixing by doing this filtering outside of core, however, here's the tl;dr:

Core stores the responses as they are from addons, api,etc. they are intended to have pages and load multiple resources in multiple states, e,g. Loading/Error state which can't be filtered in core (in case of multiple responses).

The bridge or UI parts should
1. filter data they don't need in the UI because of the data transfer between Rust and native language

2. you represent the data in UI expected format not core's raw values